container/kubernetes/onboarding/deviceonboarding.py

```diff
@@ -0,0 +1,345 @@
+    # If "activation-code" file doesn't exists
+    if '\n' in check_status:
+        check_status = check_status.splitlines()[1]
+
+    if int(check_status) == 0:
+        subprocess.call(['sh', 'environment_setup.sh'])
+
+
+def parsing_device_configdata(deploydata):
+    """ Returns multidict list from parsed onboarding
+        devices's machine and depployment config's.
+    """
+    # Parsing gateway*.yaml and creating multidict list
+    # with (deployname+nodename)==>[key] and applicationame [filteringunit,...]
+    # deploydict = defaultdict(list)  # list backed multidict
+    dictd = {}
+    deploylist = []
+    for dirs in deploydata:
+        for key in dirs:
+            # if type(dirs[key]) == dict:
+            if isinstance(dirs[key], dict):
+                for newkey, value in dirs[key].items():
+                    dictd[key+newkey] = value
+            else:
+                dictd[key] = dirs[key]
+        deploylist.append(dictd.copy())
+    return deploylist
+
+
+def process_deploy_devices_gateway(deploy_dict, deploy_type):
+    """ Method for initiating device onboarding based
+    on deployment action,deployment type and parsed
+    machine and deployment config data.
+    """
+    check_activationcode()
+    valist = []
+    # Process each device deployment configs.
+    for dict_item in deploy_dict:
+        for key in dict_item:
+            valist.append(dict_item[key])
+
+        appname = valist[0]
+        # Logic to check if Deployment Exist
+        if int(check_deployment_status(appname)) > 0:
+            if deploy_type == "create":
+                print("Deployment:", appname, "already exists",
+                      "Run Update OR Delete Deployment!!\n")
+                valist.clear()
+                continue
+        elif deploy_type in ["delete", "update"]:
+            print("Deployment:", appname, "does not exists",
+                  "Run Create Deployment First!!\n")
+            valist.clear()
+            continue
+        # Storing Nodename and Deployment Config Data
+        nodename = valist[1]
+        # Functionality needs to be implemeneted for version check
+        # toolversion = valist[2]
+        deployurl = valist[3]
+        deploydir = valist[4]
+        deploygittag = valist[5]
+
+        # TestSensor doesn't have Machine Config so check and store
+        # relevant Machine Config data.
+        machineconfig_exist = True
+        if len(valist) > 6:
+            machineurl = valist[6]
+            machinedir = valist[7]
+            machinegittag = valist[8]
+        else:
+            machineconfig_exist = False
+
+        # Clear the every Device deployment data after parse and storage.
+        valist.clear()
+
+        # Preserve application dir for "update" and "delete" deployment
+        if os.path.exists(appname) and deploy_type == "create":
+            shutil.rmtree(appname)
+
+        # Local directory doesn't exists for "update" and "delete" deployment
```

Comment:
It might also make sense to get rid of the `deploy_type` parameter, and divide this function into three `delete_deployment`, `create_deployment` ... .Then you can move out the reused parts (such as the parsing above) to further utility functions.

Reply:
Implemented and submit.